Status filter for hook log API

Yann-J opened the pull request:
## Checklist

1. [ ] If you've added code that should be tested, add tests
2. [x] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/master/CONTRIBUTING.md)
5. [ ] Write a description of your work suitable for publishing on [our forum](https://community.kobotoolbox.org/tag/release-notes)
6. [x] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes

## Description

This PR is an attempt to implement the suggestion from #3536, i.e. adding an optional `&status=status_id` query parameter to the `GET /api/v2/assets/{asset_uid}/hooks/{hook_uid}/logs/` hook log API.

## Related issues

#3536

Yann-J commented:
Community: https://community.kobotoolbox.org/t/handling-rest-service-failures/

noliveleger commented on kobo/apps/hook/views/v2/hook_log.py:
```diff
@@ -78,6 +84,10 @@ def get_queryset(self):
         # nested relations."
         queryset = queryset.select_related('hook__asset')
+
+        status = self.request.GET.get('status')
+        if status is not None:
+          queryset = queryset.filter(status=status)
```
You should add some validations on the `status` field and return a ValidationError in case user does not pass a valid status name. 

If you use `STATUSES_MAPPING`, you'd better move it to `constants.py` if your validation takes place in the serializer.

Yann-J replied:
OK the latest commit will throw on in valid statuses - but still expect the numeric status id...

noliveleger commented:
@Yann-J, Can you also add unit tests to validate your new code please?

Yann-J commented on kobo/apps/hook/views/v2/hook_log.py:
```diff
@@ -38,6 +38,12 @@ class HookLogViewSet(AssetNestedObjectViewsetMixin,
 
+    Use `status` query parameter to filter logs by numeric status:
+
+    * `status=0`: hook has failed after exhausting all retries
+    * `status=1`: hook is still pending
+    * `status=2`: hook has succeeded
```
Do you mean in the API, or in the code?
The code does not currently use the values, but I agree it should, for validation.

However, in the API, the problem is that the numeric ids are what's currently returned by the API, therefore I believe this should also be what the parameters take.
I think it would be strange to send a request with `status=failed` and then receive a response with a bunch of objects having a `{status: 0}` field...

noliveleger replied:
🤔  you are right.
This should be changed as well. (and update front-end codes too)

Yann-J commented:
> @Yann-J, Can you also add unit tests to validate your new code please?

I'll have a look but I might need guidance here... I have very limited experience in Django and even Python. I took a stab at this PR since it seemed trivial and required minimal understanding of the frameworks, and was easy to test manually end to end by just rebuilding the docker image...

Yann-J commented:
Hey @noliveleger sorry the last few commits were pretty sloppy, I've fixed the validation now, and was able to test.

I could not add new tests because I've really struggled to set up the dev environment, which has a lot of dependencies and I couldn't get them to work on my M1 Mac for some reason. So I gave up and tested end to end - which is easy through docker with `docker build .`, so I just swapped the docker image in our server.

I think it should be fairly easy for someone with a working environment to add a test in `kobo/apps/hook/tests/test_api_hook.py`, mostly copying from `test_data_submission()` but I don't understand it well enough to touch it myself...

Everything works well for me including validations:

![2022-01-19 at 14 10](https://user-images.githubusercontent.com/59641363/150127568-bdd61b0c-1686-46d7-8fe3-4575090990e4.jpg)

![2022-01-19 at 14 10](https://user-images.githubusercontent.com/59641363/150127587-563071b2-7df9-4e5d-b5ce-1337c607dbd3.jpg)

![2022-01-19 at 14 11](https://user-images.githubusercontent.com/59641363/150127598-19d7931c-88a8-40c8-a1a5-10e4b1cb203b.jpg)

![2022-01-19 at 14 11](https://user-images.githubusercontent.com/59641363/150127635-3cb6e850-72e5-4f2d-9dd7-07d771eb036c.jpg)

![2022-01-19 at 14 11](https://user-images.githubusercontent.com/59641363/150127627-77c4b105-2dbb-4f33-96a5-83f62c2654f4.jpg)

noliveleger commented:
> Hey @noliveleger sorry the last few commits were pretty sloppy, I've fixed the validation now, and was able to test.
> 
> I could not add new tests because I've really struggled to set up the dev environment, which has a lot of dependencies and I couldn't get them to work on my M1 Mac for some reason. So I gave up and tested end to end - which is easy through docker with `docker build .`, so I just swapped the docker image in our server.
> 
> I think it should be fairly easy for someone with a working environment to add a test in `kobo/apps/hook/tests/test_api_hook.py`, mostly copying from `test_data_submission()` but I don't understand it well enough to touch it myself...
> 
> Everything works well for me including validations:

Thank you for your contribution, but unfortunately, we cannot accept PRs without tests covering the new code.
If you use `kobo-install` in dev mode, you should be able to have a working environment. 
Inside the KPI container, you can run the test `pytest -vv`. 

Latest version of `beta` contains updated libraries for M1. You would need to rebuild the container though. 
From kobo-install directory:
`./run.py -cf build --force-rm --no-cache kpi`

Yann-J commented:
OK, so after much efforts I did manage to set up my local environment (even though I'm not really new at this - I'm the author of the Helm chart for KoboToolbox... how I miss that beautiful world of one-line installs!). But somehow the latest docker installers were acting up (I did eventually have to add `listen_addresses = '*'` into `postgres/shared/postgres.conf` as postgres otherwise wouldn't listen outside of its own container).

I also tried my best to figure out how the test mocking and injections seem to work. I guess I somehow managed, although clearly I don't fully understand it (what's `reverse`? what's `responses`? what's `ServiceDefinition`?).

Anyway the tests pass now, both happy path and negative (=failed hook calls), as well as input validation:

```
kobo/apps/hook/tests/test_api_hook.py::ApiHookTestCase::test_hook_log_filter_failure
kobo/apps/hook/tests/test_api_hook.py::ApiHookTestCase::test_hook_log_filter_success
kobo/apps/hook/tests/test_api_hook.py::ApiHookTestCase::test_hook_log_filter_validation
```

Hope we're all good now because I've already invested way more time on this than I was ever planning to...

Yann-J commented:
OK @noliveleger so since I'm now an expert at this, I thought I'd also through in a filter on the dates, let me know how these look!

Yann-J commented on kobo/apps/hook/tests/test_api_hook.py:
```diff
@@ -311,3 +317,160 @@ def test_payload_template_validation(self):
+           new=MockSSRFProtect._get_ip_address)
+    @responses.activate
+    def test_hook_log_filter_date(self):
+        # Create failing hook
```
Indeed!

Yann-J commented on kobo/apps/hook/tests/test_api_hook.py:
```diff
@@ -311,3 +317,160 @@ def test_payload_template_validation(self):
+        response = self.client.get(hook_log_url + '?status='+str(HOOK_LOG_FAILED), format='json')
+        self.assertEqual(response.data.get('count'), 0)
+
+        # Test bad argument
+        response = self.client.get(hook_log_url + '?status=abc', format='json')
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
👍

Yann-J commented on kobo/apps/hook/tests/test_api_hook.py:
```diff
@@ -311,3 +317,160 @@ def test_payload_template_validation(self):
+            'parent_lookup_hook': hook.uid,
+        })
+
+        now = datetime.utcnow()
```
OK I actually found out about `kpi.utils.datetime.several_minutes_from_now` so just used that.

Yann-J commented on kobo/apps/hook/tests/test_api_hook.py:
```diff
@@ -311,3 +317,160 @@ def test_payload_template_validation(self):
+        response = self.client.get(hook_log_url + '?end='+str(now - delta), format='json')
+        self.assertEqual(response.data.get('count'), 0)
+
+        # There should be no log around now when expressed in a different time zone
+        response = self.client.get(hook_log_url + '?start='+str(now - delta) + tzoffset + '&end='+str(now + delta) + tzoffset, format='json')
+        self.assertEqual(response.data.get('count'), 0)
```
Passing the offset (NOT the time zone!) is part of the ISO 8601 standard and actually parsed by `datetime.fromisoformat`, so there's really no reason not to accept those (in fact it might be more work) - even if all the storage and calculations of course remain in UTC.

Yann-J commented on kobo/apps/hook/tests/test_api_hook.py:
```diff
@@ -311,3 +317,160 @@ def test_payload_template_validation(self):
+        tzoffset = '-02:00'
+
+        # There should be a success log around now
+        response = self.client.get(hook_log_url + '?start='+str(now - delta) + '&end='+str(now + delta), format='json')
```
👍

Yann-J commented on kobo/apps/hook/views/v2/hook_log.py:
```diff
@@ -1,17 +1,25 @@
+    HOOK_LOG_SUCCESS,
+    KOBO_INTERNAL_ERROR_STATUS_CODE,
+)
+VALID_STATUSES = [HOOK_LOG_FAILED, HOOK_LOG_PENDING, HOOK_LOG_SUCCESS]
```
I really think human-readable statuses are outside the scope of this PR, which only adds filters and does not deal with the API response format.
Besides, that would be a breaking API change so I think you would need to plan this carefully...

Yann-J commented on kobo/apps/hook/views/v2/hook_log.py:
```diff
@@ -78,6 +96,43 @@ def get_queryset(self):
         queryset = queryset.select_related('hook__asset')
+
+        # Filter on status
+        status = self.request.GET.get('status')
+        if status is not None:
+            if status not in map(str,VALID_STATUSES):
+                raise serializers.ValidationError(
+                    {'status': _('Value must be one of: ' +
+                                 ', '.join(map(str,VALID_STATUSES)))}
+                )
+            else:
+                queryset = queryset.filter(status=status)
+
+        # Filter on date range
+        start = self.request.GET.get('start')
+        if start is not None:
+          try:
+            start_date = datetime.fromisoformat(start)
+            if not start_date.tzname():
+                start_date=start_date.replace(tzinfo=timezone.utc)
+            queryset = queryset.filter(date_modified__gte=start_date)
+          except ValueError:
+            raise serializers.ValidationError(
+                {'start': _('Value must be a valid ISO-8601 date')}
+            )
+
+        end = self.request.GET.get('end')
+        if end is not None:
+          try:
+            end_date = datetime.fromisoformat(end)
+            if not end_date.tzname():
+                end_date=end_date.replace(tzinfo=timezone.utc)
+            queryset = queryset.filter(date_modified__lt=end_date)
+          except ValueError:
+            raise serializers.ValidationError(
+                {'end': _('Value must be a valid ISO-8601 date')}
+            )
```
I'm not really up for it... I think I've already gone way beyond what can be reasonably expected of a casual community contributor. I've made it clear that I was a noob in Python/Django and yet was still encouraged to "give it a try", but then got sucked into having to learn a mostly undocumented set of standards (why isn't pep-8 compliance enforced by a linter instead of relying on manual reviews?), dev environment setup, test framework, that I think you will agree are a pretty steep learning curve.

If you hope to get more community contributions, I really think you need to invest into some dev documentation and toolchain setup...

Yann-J commented:
Hey @noliveleger any news on this one? I believe I've addressed most comments.
I also noticed you changed the target branch, but it had conflicts, which I just resolved now.

Yann-J commented:
@noliveleger are we good to go?

jnm commented:
Hi @Yann-J, thanks for your work on this so far. I agree that the amount of effort you have invested here is quite significant for a casual contributor. We will add the remaining work (as described in @noliveleger's review) to our core developers' backlog. Unfortunately, this PR cannot be merged until that work is complete.

bufke commented:
Can continue discussion on https://github.com/kobotoolbox/kpi/pull/3737

noliveleger commented:
> The breaking API change for status is a big one. What's our policy on breaking api changes? Do we need to wait for a major release? I agree it's unfortunate that we are using a number that really is more of a DB optimization artifact and should never have been exposed on an api. I'm tempted to say hold off on that change as it's own PR. We do expose status_str already. It would be kind of gross to filter on status_str but I'll through that out.

We usually keep the old code for a while and let the users know in the release notes. 
FYI: `/api/v1/` has been deprecated for a while but it is still alive.
